File: sdk/py/splendor_actor/types/card.py
from typing import List, Dict, Union
from enum import Enum

from .color import Color, ColorVec
from .tier import Tier


class Card:
    """
    represent the development cards
    """

    tier: Tier
    bonus: Color
    points: int
    requires: ColorVec

    def __init__(self, tier: int, bonus: str, points: int, requires: List[int]):
        self.tier = Tier(tier)
        self.bonus = Color.from_str(bonus)
        self.points = points
        self.requires = ColorVec(requires)


class CardPool:
    """
    represent the development cards pool
    """

    remaining: List[int]
    revealed: List[List[Card]]

    def __init__(self, remaining: List[int], revealed: List[List[Dict]]):
        self.remaining = remaining
        self.revealed = [[Card(**card) for card in tier] for tier in revealed]


class ReservedCardType(Enum):
    """
    Enum for the reserved card type.
    Reserved cards can be invisible if they are reserved from pool.
    """

    Visible = 0
    Invisible = 1

    def from_str(s: str) -> "ReservedCardType":
        """
        Convert a string to a ReservedCardType enum
        """
        s = s.lower()
        if s == "visible":
            return ReservedCardType.Visible
        elif s == "invisible":
            return ReservedCardType.Invisible
        else:
            raise ValueError(f"Invalid reserved card type: {s}")

    def __str__(self):
        return self.name

    def __repr__(self):
        return self.name


class CardView:
    """
    represent the reserved cards
    """

    type: ReservedCardType
    view: Union[Card, Tier]

    def __init__(self, type: str, view: Union[Dict, int]):
        self.type = ReservedCardType.from_str(type)
        if self.type == ReservedCardType.Visible:
            self.view = Card(**view)
        else:
            self.view = Tier(view)


class DevelopmentCards:
    """
    represent the development cards of a player
    """

    points: int
    bonus: ColorVec
    inner: List[List[Card]]

    def __init__(self, points: int, bonus: List[int], inner: List[List[Dict]]):
        self.points = points
        self.bonus = ColorVec(bonus)
        self.inner = [[Card(**card) for card in tier] for tier in inner]

File: sdk/go/pkg/types/card.go
package types

import (
	"encoding/json"

	log "github.com/sirupsen/logrus"
)

// Card represents a card in the game
type Card struct {
	// The tier of the card
	Tier Tier `json:"tier"`
	// The color of the card
	Color Color `json:"color"`
	// The points of the card
	Points uint8 `json:"points"`
	// The cost of the card
	Requires ColorVec `json:"requires"`
}

// The card pool
type CardPool struct {
	// The remaining cards in each tier
	Remaining [3]uint8 `json:"remaining"`
	// The revealed cards in each tier
	Revealed [3][]Card `json:"revealed"`
}

const (
	VISIBLE_RESERVED_CARD   string = "visible"
	INVISIBLE_RESERVED_CARD string = "invisible"
)

// CardView represents a card that can be visible or not
type CardView struct {
	Type string      `json:"type"`
	View interface{} `json:"view"`
}

// UnwrapCard unwraps the card from the view, panics if the card is not visible
func (c *CardView) UnwrapCard() Card {
	if c.Type == VISIBLE_RESERVED_CARD {
		cardMashal, _ := json.Marshal(c.View)
		var card Card
		json.Unmarshal(cardMashal, &card)
		return card
	}
	log.Panic("cannot unwrap invisible card")
	return Card{} // unreachable
}

// DevelopmentCards represents the development cards of a player
type DevelopmentCards struct {
	// The total points of the development cards
	Points uint8 `json:"points"`
	// The total bonus tokens of the development cards
	Bonus ColorVec `json:"bonus"`
	// The cards grouped by color
	Inner [6][]Card `json:"inner"`
}
